dali/operators/reader/loader/numpy_loader_gpu.cc

```diff
@@ -0,0 +1,225 @@
+  DALI_ENFORCE((header_len + 10) % 16 == 0,
+               "Error extracting header length.");
+
+  // read header: the offset is a magic number
+  int64 offset = (6+1+1+2);
+  // the header_len can be 4GiB according to the NPYv2 file format
+  // specification: https://numpy.org/neps/nep-0001-npy-format.html
+  // while this allocation could be sizable, it is performed on the host.
+  token.resize(header_len+1);
+  file->Seek(offset);
+  nread = file->ReadCPU(token.data(), header_len);
+  DALI_ENFORCE(nread == header_len, "Can not read header.");
+  token[header_len] = '\0';
+  header = std::string(reinterpret_cast<char*>(token.data()));
+  DALI_ENFORCE(header.find_first_of("{") != std::string::npos, "Header is corrupted.");
+  offset += header_len;
+
+  // store the file offset in the parse target
+  target.data_offset = offset;
+
+  // prepare file for later reads
+  file->Seek(offset);
+
+  // extract dictionary info from header
+  std::smatch header_match;
+  DALI_ENFORCE(std::regex_search(header, header_match, header_regex_),
+               "Can not parse header.");
+
+  // now extract header information
+  // type
+  std::string typestring = header_match[1].str();
+
+  // < means LE, | means N/A, = means native. In all those cases, we can read
+  bool little_endian =
+    (typestring[0] == '<' || typestring[0] == '|' || typestring[0] == '=');
+  DALI_ENFORCE(little_endian,
+    "Big Endian files are not supported.");
+
+  std::string tid = typestring.substr(1);
+  // get type in a safe way
+  target.type_info = TypeFromNumpyStr(tid);
+
+  // check for data order
+  if (header_match[2].str() == "False")
+    target.fortran_order = false;
+  else
+    target.fortran_order = true;
+
+  // set sizes
+  std::string shapestring = header_match[3].str();
+  std::regex shape_regex{R"(,+)"};  // split on comma
+  std::sregex_token_iterator it{shapestring.begin(), shapestring.end(), shape_regex, -1};
+  std::vector<std::string> shapevec{it, {}};
+
+  // if shapevec size is 1 and shapevec[0] is the empty string,
+  // the array is actually a scalar/singleton (denoted as ())
+  // and thus the size needs to be set to one:
+  if ( (shapevec.size() == 1) && (shapevec[0] == "") ) shapevec[0] = "1";
+
+  // determine shapes
+  size_t shapesize = shapevec.size();
+  target.shape.resize(shapesize);
+  // cheapest thing to do is to define the tensor in an reversed way
+  if (target.fortran_order) {
+    for (size_t i = 0; i < shapesize; ++i)
+      target.shape[i] = static_cast<int64_t>(stoi(shapevec[shapesize-i-1]));
+  } else {
+    for (size_t i = 0; i < shapesize; ++i)
+      target.shape[i] = static_cast<int64_t>(stoi(shapevec[i]));
+  }
+
+  return file;
+}
+
+// register tensor
+void NumpyLoaderGPU::RegisterTensor(void *buffer, size_t total_size) {
+  if (register_buffers_) {
+    // get raw pointer
+    auto dptr = static_cast<uint8_t*>(buffer);
+    std::unique_lock<std::mutex> reg_lock(reg_mutex_);
+    auto iter = reg_buff_.find(dptr);
```

Comment:
```suggestion
    auto iter = reg_buff_.find(dptr);
    reg_loc.unlock();
```
and you can remove the unlock inside the if/else (and remove the else completely)

Reply:
Done